File: app/src/com/google/android/stardroid/StardroidApplication.java
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid;

import com.google.android.stardroid.activities.DynamicStarMapActivity;
import com.google.android.stardroid.control.AstronomerModel;
import com.google.android.stardroid.control.AstronomerModelImpl;
import com.google.android.stardroid.control.ZeroMagneticDeclinationCalculator;
import com.google.android.stardroid.layers.EclipticLayer;
import com.google.android.stardroid.layers.GridLayer;
import com.google.android.stardroid.layers.HorizonLayer;
import com.google.android.stardroid.layers.LayerManager;
import com.google.android.stardroid.layers.MeteorShowerLayer;
import com.google.android.stardroid.layers.NewConstellationsLayer;
import com.google.android.stardroid.layers.NewMessierLayer;
import com.google.android.stardroid.layers.NewStarsLayer;
import com.google.android.stardroid.layers.PlanetsLayer;
import com.google.android.stardroid.layers.SkyGradientLayer;
import com.google.android.stardroid.util.Analytics;
import com.google.android.stardroid.util.Analytics.Slice;
import com.google.android.stardroid.util.MiscUtil;
import com.google.android.stardroid.util.OsVersions;
import com.google.android.stardroid.util.PreferenceChangeAnalyticsTracker;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.PackageManager.NameNotFoundException;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.preference.PreferenceManager;
import android.util.Log;

import java.util.Calendar;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.ScheduledThreadPoolExecutor;

/**
 * The main Stardroid Application class.
 *
 * @author John Taylor
 */
public class StardroidApplication extends Application {
  private static final String TAG = MiscUtil.getTag(StardroidApplication.class);
  private static final String PREVIOUS_APP_VERSION_PREF = "previous_app_version";
  private static final String NONE = "Clean install";
  private static final String UNKNOWN = "Unknown previous version";
  // The Application class is a singleton, so treat it as such, with static
  // fields.  This is necessary so that the content provider can access the
  // things it needs; there seems to be no easy way for a ContentProvider
  // to access its Application object.
  private static AstronomerModel model;
  private static LayerManager layerManager;
  private static ExecutorService backgroundExecutor;

  // We need to maintain references to this object to keep it from
  // getting gc'd.
  private final PreferenceChangeAnalyticsTracker preferenceChangeAnalyticsTracker =
      new PreferenceChangeAnalyticsTracker(Analytics.getInstance(this));


  @Override
  public void onCreate() {
    Log.d(TAG, "StardroidApplication: onCreate");
    super.onCreate();

    Log.i(TAG, "OS Version: " + android.os.Build.VERSION.RELEASE
        + "(" + android.os.Build.VERSION.SDK + ")");
    String versionName = getVersionName();
    Log.i(TAG, "Sky Map version " + versionName + " build " + getVersion());
    backgroundExecutor = new ScheduledThreadPoolExecutor(1);
    // This populates the default values from the preferences XML file. See
    // {@link DefaultValues} for more details.
    PreferenceManager.setDefaultValues(this, R.xml.preference_screen, false);

    AssetManager assetManager = this.getAssets();
    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
    Resources resources = this.getResources();
    // Start the LayerManager initializing
    getLayerManager(assetManager, preferences, resources, this);

    setUpAnalytics(versionName, preferences);

    Log.d(TAG, "StardroidApplication: -onCreate");
  }

  private void setUpAnalytics(String versionName, SharedPreferences preferences) {
    Analytics analytics = Analytics.getInstance(this);
    analytics.setProductVersion(versionName);
    analytics.setCustomVar(Slice.ANDROID_OS, Integer.toString(OsVersions.version()));
    analytics.setCustomVar(Slice.SKYMAP_VERSION, versionName);
    analytics.setCustomVar(Slice.DEVICE_NAME, android.os.Build.MODEL);
    analytics.setEnabled(preferences.getBoolean(Analytics.PREF_KEY, true));
    analytics.trackPageView(Analytics.APPLICATION_CREATE);

    String previousVersion = preferences.getString(PREVIOUS_APP_VERSION_PREF, NONE);
    if (previousVersion.equals(NONE)) {
      // It's possible a previous version exists, it's just that it wasn't a recent enough
      // version to have set PREVIOUS_APP_VERSION_PREF.  If so, we should see that the TOS
      // have been accepted.
      if (preferences.contains(DynamicStarMapActivity.READ_TOS_PREF)) {
        previousVersion = UNKNOWN;
      }
    }
    preferences.edit().putString(PREVIOUS_APP_VERSION_PREF, versionName).commit();
    if (!previousVersion.equals(versionName)) {
      // It's either an upgrade or a new installation
      Log.d(TAG, "New installation: version " + versionName);
      analytics.trackEvent(Analytics.INSTALL_CATEGORY, Analytics.INSTALL_EVENT + versionName,
          Analytics.PREVIOUS_VERSION + previousVersion, 1);
    }

    // It will be interesting to see *when* people use Sky Map.
    analytics.trackEvent(
        Analytics.GENERAL_CATEGORY, Analytics.START_HOUR,
        Integer.toString(Calendar.getInstance().get(Calendar.HOUR_OF_DAY)) + 'h', 0);

    preferences.registerOnSharedPreferenceChangeListener(preferenceChangeAnalyticsTracker);
  }

  @Override
  public void onTerminate() {
    super.onTerminate();
    Analytics.getInstance(this).setEnabled(false);
  }

  /**
   * Returns the version string for Sky Map.
   */
  public String getVersionName() {
    PackageManager packageManager = getPackageManager();
    try {
      PackageInfo info = packageManager.getPackageInfo(this.getPackageName(), 0);
      return info.versionName;
    } catch (NameNotFoundException e) {
      Log.e(TAG, "Unable to obtain package info");
      return "Unknown";
    }
  }

  /**
   * Returns the build number for Sky Map.
   */
  public int getVersion() {
    PackageManager packageManager = getPackageManager();
    try {
      PackageInfo info = packageManager.getPackageInfo(this.getPackageName(), 0);
      return info.versionCode;
    } catch (NameNotFoundException e) {
      Log.e(TAG, "Unable to obtain package info");
      return -1;
    }
  }

  /**
   * Get the catalog.
   * This should return relatively quickly, with the catalogs initializing
   * themselves on background threads.
   */
  public static synchronized LayerManager getLayerManager(AssetManager assetManager,
                                                          SharedPreferences preferences,
                                                          Resources resources,
                                                          Context context) {
    if (layerManager == null) {
      Log.i(TAG, "Initializing LayerManager");
      layerManager = new LayerManager(preferences, getModel());
      layerManager.addLayer(new NewStarsLayer(assetManager, resources));
      layerManager.addLayer(new NewMessierLayer(assetManager, resources));
      layerManager.addLayer(new NewConstellationsLayer(assetManager, resources));
      layerManager.addLayer(new PlanetsLayer(getModel(), resources, preferences));
      layerManager.addLayer(new MeteorShowerLayer(getModel(), resources));
      layerManager.addLayer(new GridLayer(resources, 24, 19));
      layerManager.addLayer(new HorizonLayer(getModel(), resources));
      layerManager.addLayer(new EclipticLayer(resources));
      layerManager.addLayer(new SkyGradientLayer(getModel(), resources));
      // layerManager.addLayer(new IssLayer(resources, getModel()));
      layerManager.initialize();
    } else {
      Log.i(TAG, "LayerManager already initialized.");
    }
    return layerManager;
  }

  /**
   * Return the model.
   */
  public static synchronized AstronomerModel getModel() {
    if (model == null) {
      model = new AstronomerModelImpl(new ZeroMagneticDeclinationCalculator());
    }
    return model;
  }

  /**
   * Schedules this runnable to run as soon as possible on a background
   * thread.
   *
   * @param runnable
   */
  // TODO(johntaylor): the idea, and I'm not sure yet whether it's a good one,
  // is to centralize the management of background threads so we don't have
  // them scattered all over the app.  We can then control how many threads
  // are spawned, perhaps having a VIP service for extra important runnables
  // that we'd prefer not to queue, as well as providing convenience functions
  // to facilitate callbacks on the UI thread.
  public static void runInBackground(Runnable runnable) {
    backgroundExecutor.submit(runnable);
  }
}

File: app/src/main/java/com/google/android/stardroid/StardroidApplication.kt
// Copyright 2008 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.google.android.stardroid

import android.app.Application
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import androidx.core.content.pm.PackageInfoCompat
import androidx.preference.PreferenceManager
import com.google.android.stardroid.layers.LayerManager
import com.google.android.stardroid.util.AnalyticsInterface
import com.google.android.stardroid.util.MiscUtil.getTag
import com.google.android.stardroid.util.PreferenceChangeAnalyticsTracker
import com.google.android.stardroid.views.PreferencesButton
import java.util.*
import javax.inject.Inject

/**
 * The main Stardroid Application class.
 *
 * @author John Taylor
 */
class StardroidApplication : Application() {
  @Inject
  lateinit var preferences: SharedPreferences

  // We keep a reference to this just to start it initializing.
  @Inject
  lateinit var layerManager: LayerManager

  @Inject
  lateinit var analytics: AnalyticsInterface

  @Inject
  @JvmField
  var sensorManager: SensorManager? = null

  // We need to maintain references to this object to keep it from
  // getting gc'd.
  @Inject
  lateinit var preferenceChangeAnalyticsTracker: PreferenceChangeAnalyticsTracker

  val applicationComponent: ApplicationComponent = DaggerApplicationComponent.builder()
    .applicationModule(ApplicationModule(this))
    .build()

  override fun onCreate() {
    Log.d(TAG, "StardroidApplication: onCreate")
    super.onCreate()
    applicationComponent.inject(this)
    Log.i(
      TAG, "OS Version: " + Build.VERSION.RELEASE
          + "(" + Build.VERSION.SDK_INT + ")"
    )
    val versionName = versionName
    Log.i(TAG, "Sky Map version $versionName build $version")

    // This populates the default values from the preferences XML file. See
    // {@link DefaultValues} for more details.
    PreferenceManager.setDefaultValues(this, R.xml.preference_screen, false)
    setUpAnalytics(versionName)
    performFeatureCheck()
    Log.d(TAG, "StardroidApplication: -onCreate")
  }

  private fun setUpAnalytics(versionName: String) {
    analytics.setEnabled(preferences.getBoolean(AnalyticsInterface.PREF_KEY, true))

    // Ugly hack since this isn't injectable
    PreferencesButton.setAnalytics(analytics)
    var previousVersion = preferences.getString(PREVIOUS_APP_VERSION_PREF, NONE)
    var newUser = false
    if (previousVersion == NONE) {
      // It's possible a previous version exists, it's just that it wasn't a recent enough
      // version to have set PREVIOUS_APP_VERSION_PREF.  If so, we should see that the TOS
      // have been accepted.
      val oldPreviousVersionKey = "read_tos"
      if (preferences.contains(oldPreviousVersionKey)) {
        previousVersion = UNKNOWN
      } else {
        // Best guess that this is the first every run of a new user.
        // Could also be someone with a new device.
        newUser = true
      }
    }
    analytics.setUserProperty(AnalyticsInterface.NEW_USER, java.lang.Boolean.toString(newUser))
    preferences.edit().putString(PREVIOUS_APP_VERSION_PREF, versionName).apply()
    if (previousVersion != versionName) {
      // It's either an upgrade or a new installation
      Log.d(TAG, "New installation: version $versionName")
      // No need to track any more - it's automatic in Firebase.
    }

    // It will be interesting to see *when* people use Sky Map.
    val b = Bundle()
    b.putInt(AnalyticsInterface.START_EVENT_HOUR, Calendar.getInstance()[Calendar.HOUR_OF_DAY])
    analytics.trackEvent(AnalyticsInterface.START_EVENT, b)
    preferences.registerOnSharedPreferenceChangeListener(preferenceChangeAnalyticsTracker)
  }

  override fun onTerminate() {
    super.onTerminate()
    analytics.setEnabled(false)
  }// TODO(jontayler): update to use the info created by gradle.

  /**
   * Returns the version string for Sky Map.
   */
  val versionName: String
    get() {
      // TODO(jontayler): update to use the info created by gradle.
      val packageManager = packageManager
      return try {
        val info = packageManager.getPackageInfo(this.packageName, 0)
        info.versionName
      } catch (e: PackageManager.NameNotFoundException) {
        Log.e(TAG, "Unable to obtain package info")
        "Unknown"
      }
    }

  /**
   * Returns the build number for Sky Map.
   */
  val version: Long
    get() {
      val packageManager = packageManager
      return try {
        val info = packageManager.getPackageInfo(this.packageName, 0)
        PackageInfoCompat.getLongVersionCode(info)
      } catch (e: PackageManager.NameNotFoundException) {
        Log.e(TAG, "Unable to obtain package info")
        -1
      }
    }

  /**
   * Check what features are available to this phone and report back to analytics
   * so we can judge when to add/drop support.
   */
  private fun performFeatureCheck() {
    if (sensorManager == null) {
      Log.e(TAG, "No sensor manager")
      analytics.setUserProperty(
        AnalyticsInterface.DEVICE_SENSORS, AnalyticsInterface.DEVICE_SENSORS_NONE
      )
      return
    }
    // Reported available sensors
    val reportedSensors: MutableList<String> = ArrayList()
    if (hasDefaultSensor(Sensor.TYPE_ACCELEROMETER)) {
      reportedSensors.add(AnalyticsInterface.DEVICE_SENSORS_ACCELEROMETER)
    }
    if (hasDefaultSensor(Sensor.TYPE_GYROSCOPE)) {
      reportedSensors.add(AnalyticsInterface.DEVICE_SENSORS_GYRO)
    }
    if (hasDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)) {
      reportedSensors.add(AnalyticsInterface.DEVICE_SENSORS_MAGNETIC)
    }
    if (hasDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)) {
      reportedSensors.add(AnalyticsInterface.DEVICE_SENSORS_ROTATION)
    }

    // TODO: Change to String.join once we're at API > 26
    analytics.setUserProperty(
      AnalyticsInterface.DEVICE_SENSORS, TextUtils.join("|", reportedSensors)
    )

    // Check for a particularly strange combo - it would be weird to have a rotation sensor
    // but no accelerometer or magnetic field sensor
    var hasRotationSensor = false
    if (hasDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)) {
      if (hasDefaultSensor(Sensor.TYPE_ACCELEROMETER) && hasDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)
        && hasDefaultSensor(Sensor.TYPE_GYROSCOPE)
      ) {
        hasRotationSensor = true
      } else if (hasDefaultSensor(Sensor.TYPE_ACCELEROMETER) && hasDefaultSensor(
          Sensor.TYPE_MAGNETIC_FIELD
        )
      ) {
        // Even though it allegedly has the rotation vector sensor too many gyro-less phones
        // lie about this, so put these devices on the 'classic' sensor code for now.
        hasRotationSensor = false
      }
    }

    // Enable Gyro if available and user hasn't already disabled it.
    if (!preferences.contains(ApplicationConstants.SHARED_PREFERENCE_DISABLE_GYRO)) {
      preferences.edit().putBoolean(
        ApplicationConstants.SHARED_PREFERENCE_DISABLE_GYRO, !hasRotationSensor
      ).apply()
    }

    // Lastly a dump of all the sensors.
    Log.d(TAG, "All sensors:")
    val allSensors = sensorManager?.getSensorList(Sensor.TYPE_ALL)
    val sensorTypes: MutableSet<String> = HashSet()
    for (sensor in allSensors ?: emptyList()) {
      Log.i(TAG, sensor.name)
      sensorTypes.add(getSafeNameForSensor(sensor))
    }
    Log.d(TAG, "All sensors summary:")
    for (sensorType in sensorTypes) {
      Log.i(TAG, sensorType)
    }
  }

  private fun hasDefaultSensor(sensorType: Int): Boolean {
    val sensor = sensorManager?.getDefaultSensor(sensorType) ?: return false
    val dummy: SensorEventListener = object : SensorEventListener {
      override fun onSensorChanged(event: SensorEvent) {
        // Nothing
      }

      override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        // Nothing
      }
    }
    val success = sensorManager?.registerListener(
      dummy, sensor, SensorManager.SENSOR_DELAY_UI
    ) ?: false
    if (!success) {
      analytics.setUserProperty(AnalyticsInterface.SENSOR_LIAR, "true")
    }
    sensorManager?.unregisterListener(dummy)
    return success
  }

  companion object {
    private val TAG = getTag(StardroidApplication::class.java)
    private const val PREVIOUS_APP_VERSION_PREF = "previous_app_version"
    private const val NONE = "Clean install"
    private const val UNKNOWN = "Unknown previous version"

    /**
     * Returns either the name of the sensor or a string version of the sensor type id, depending
     * on the supported OS level along with some context.
     */
    fun getSafeNameForSensor(sensor: Sensor) = "Sensor type: ${sensor.stringType}: ${sensor.type}"
  }
}